### PR TITLE
feat(tailwind): Smaller bundle size

### LINKED ISSUE #1104: Hooks + Context not working within `Tailwind` component
The new Tailwind component makes it quite impossible to use hooks or context w/ react email.

As I'm using react-i18next for i18n, I'm using the `useTranslation` hook. So I'm not having a dynamic use case either, just using available tools.

Also, I'm using context to pass around some shared data without using prop drilling.

My workaround till now was having `Tailwind` as my root, so it won't interfere w/ context etc.

However, this won't work with the rewrite, as it's now required to have the `<head>` as a direct descendant of `<Tailwind>`.

The hooks issue was also mentioned in #1021, and also what's the probable cause:

>> @gabrielmfern that does seem a little weird to me, as I do expect hooks to run for other "static" use cases like static site generation or server-side rendering, especially as async server side rendering is becoming more of a thing, and indeed hooks run fine with `renderToStaticMarkup`. React contexts also work fine in static rendering, the issue is with how the `tailwind` function traverses the component tree and calls components as plain functions: https://github.com/resendlabs/react-email/blob/2374d0d099360c63fe7427febb239776c3bd6b3b/packages/tailwind/src/tailwind.tsx#L95C32-L95C32.
>> 
>> For now my workaround is using `<MyReactContext.Consumer>{context => ...}</MyReactContext.Consumer>` instead of the hook.

> That's interesting, then I think this is an actual issue with the new Tailwind version, can you open an issue on this so we can reference it on a PR and discuss it there?

> _Originally posted by @gabrielmfern in https://github.com/resend/react-email/issues/1021#issuecomment-1850747538_

### PR DESCRIPTION
**Pull request by gabrielmfern**
## Background

After the updates to the Tailwind component that came with performance improvements
we had a lot more environments to consider while trying to run `tailwindcss`
under the hood for generating the styles from classes.

These environment constraints, like being able to run on the browser and on other environments,
forced us into having to polyfill certain node modules — that both `postcss` and `tailwindcss`
use under the hood — while bundling even though these code paths were never really
hit when running. 

The reason these code paths were not completely tree shaken out of the built code was because 
we called `tailwindcss` directly as a plugin on `postcss`.

> Side note: This PR was first intended for #1104 but ended up into a better implementation of #1124

## Why?

Polyfilling was the temporary solution for this, but it is not optimal in the least. This caused
issues like:

- #1101
- #1244
- #1105

And ended up also being a burden to maintain and debug.

## The solution

The best way I've found to solve these issues is by calling out the internal Tailwind functions
that do what they need to do on `postcss` data structures. This also allows us to substitute a bit of
our logic, which was previously purely done with regexes, with `postcss`'s utilities which makes it much easier to read
and maintain.

The reason I choose to call out the internal Tailwind functions was because the code that was Node specific
was really just around finding the configuration or checking if a certain thing was installed, and
this would really only run when it was called as a `postcss` plugin.

The way I do call out the internal Tailwind functions is heavily inspired by the way they do it
themselves on the Tailwind VS Code extension/LSP Server.

## Other changes that make this a refactor

Along with this, I also had to change up a few things. The first and biggest one was applying
the non-inlinable styles — currently media queries, i.e. `sm:...` — after our first pass
on the React tree which is for inlining styles. This is because we are now generating the styles
for each element's classes on-the-fly instead of at once (which is still performant the way it is
implemented), so we need to aggregate all the non-inlinable classes and their CSS to add to the `<head>` 
later.

The second change should be mostly internal, (and should also make it easier for others to contribute),
which is a new `mapReactTree` utility that goes through the whole tree and separates the concerns
about when to go recursive on the elements or not. It will also make it easier to actually look at #1104 once
we get there.

## Added tests for ensured stability

As I mentioned, we had to consider a bunch of different environments and testing them manually
was really time-consuming, so, borrowed from #1124, I implemented an integration tested powered by [yalc](https://github.com/wclr/yalc) that allows us to make sure that the component will work well with NextJS's build process.

You can find this test inside `packages/tailwind/integrations/_tests/nextjs.spec.ts`. This PR also adds a test for `Vite` as well.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -28,7 +28,6 @@
   "pnpm": {
     "patchedDependencies": {
       "postcss-css-variables@0.19.0": "patches/postcss-css-variables@0.19.0.patch",
-      "process@0.11.10": "patches/process@0.11.10.patch",
```
Comment (gabrielmfern) [inline package.json:31]: This package was one that we used as a polyfill that isn't needed and isn't required to be patched anymore!

File: packages/tailwind/src/tailwind.tsx [hunk 2]
```diff
@@ -12,133 +13,91 @@ export interface TailwindProps {
...
-      const renderedComponent = component({
-        ...element.props,
-        ...propsToOverwrite,
-      });
-
-      if (React.isValidElement<EmailElementProps>(renderedComponent)) {
-        return processElement(renderedComponent);
+  if (hasNonInlineStylesToApply) {
+    let hasAppliedNonInlineStyles = false as boolean;
+    mappedChildren = mapReactTree(mappedChildren, (node) => {
```
Comment (gabrielmfern) [inline packages/tailwind/src/tailwind.tsx:57]: Following the same idea from what we merged recently that allows the `<head>` element to be at any depth, this maps the entire React tree to find it.

File: packages/tailwind/src/tailwind.spec.tsx [hunk 3]
```diff
@@ -22,6 +22,34 @@ describe("Tailwind component", () => {
     });
   });
 
+  // test("with React context and custom components", () => {
```
Comment (gabrielmfern) [inline packages/tailwind/src/tailwind.spec.tsx:25]: A unit test I kept commented out as useful for when we get to #1104.

File: packages/tailwind/src/tailwind.spec.tsx [hunk 4]
```diff
@@ -33,7 +61,7 @@ describe("Tailwind component", () => {
...
+      `"<a style="line-height:1.25rem;text-decoration:none;display:inline-block;max-width:100%;padding-left:0.75rem;padding-right:0.75rem;padding-top:0.5rem;padding-bottom:0.5rem;margin-top:2rem;font-size:0.875rem;color:rgb(229,231,235);background-color:rgb(37,99,235);border-radius:0.375rem;padding:8px 12px 8px 12px" target="_blank"><span><!--[if mso]><i style="letter-spacing: 12px;mso-font-width:-100%;mso-text-raise:12" hidden>&nbsp;</i><![endif]--></span><span style="max-width:100%;display:inline-block;line-height:120%;mso-padding-alt:0px;mso-text-raise:6px">Testing button</span><span><!--[if mso]><i style="letter-spacing: 12px;mso-font-width:-100%" hidden>&nbsp;</i><![endif]--></span></a>Testing"`,
```
Comment (gabrielmfern) [inline packages/tailwind/src/tailwind.spec.tsx:64]: These changes on the tests' snapshots are because of the update to a newer vitest version where escaping `"` isn't needed anymore.

**Comment by VaniaPopovic** [comment]
Hi, is there any way we can get a patch for pnpm for this one? I would like to test whether it actually fixes one of the issues that we are not 100% sure this fixes :)

**Comment by gabrielmfern** [comment]
@VaniaPopovic Since this is going to be quite a big change, I recommend you rather download the repo, build it and set the version of the `@react-email/tailwind` as a path to the cloned repo. So:

```diff
- "@react-email/tailwind": "0.0.15"
+ "@react-email/tailwind": "/path/to/clone/packages/tailwind"
```

This is going to be mostly representative of how it will run when installing it from `npm`. If you want something even more accurate, you can check out [yalc](https://github.com/wclr/yalc). 

Also, let me know what other kinds of issues this fixes!

**Comment by teovillanueva** [comment]
Hey! We would also like to try if these changes fix our issue (#1101). Is there an estimated release date for this? Thanks!

**Comment by gabrielmfern** [comment]
Hey @teovillanueva, can't give an ETA for now, but you can enable notifications and once we do review and merge this you will be notified.

**Comment by Xexr** [comment]
Fwiw, I've been using a fork of this to get compatibility with nextJS 14.3.0.

It's been working well so far.

**Comment by gabrielmfern** [comment]
@Xexr is it jsx-email?

**Comment by Xexr** [comment]
@gabrielmfern Yes its jsx-email, but I spoke to soon.

I actually now get the error:

```
 TypeError: renderToStream is not a function
```

The original issue I had is this: https://github.com/resend/react-email/issues/1413

So I've been trying to run the tailwind package from this in the interim so that I can use React19 (via NextJS 14.3.0)

It does now build correctly, but when sending an email I now get the above error.

**Comment by gabrielmfern** [comment]
@Xexr Hmm, thinking this might be fixed by #1443 since Next's canaries are using beta React, it also overrides the `react-dom` version you have installed, meaning it's basically using React 19. But, Resend will also need to upgrade right after that is released to fix it as well.

**Comment by Xexr** [comment]
Thanks @gabrielmfern 

#1443 allowed me to progress past that error.

I merged it in to react-email render and then used that in a fork of resend with react/react-dom pegged to the 19 beta.

However it now appears to get stuck in a loop and doesn't finish rendering.

**Comment by gabrielmfern** [comment]
> However it now appears to get stuck in a loop and doesn't finish rendering.

@Xexr We do now have #1443 released on the latest `@react-email/render` and `@react-email/components`, can you make a repro with them? All the situations I tried with this change to the render still worked.

Also, are you rendering this on the browser or on the server? It uses a different `react-dom` function depending on the environment, so this could be a clue.

**Comment by Xexr** [comment]
Hi @gabrielmfern, thanks for this.

This was rendering on the server and it was resend that was stuck in a loop specifically.

I've moved to a different architecture to get around this with react18 dependent packages on a separate server pegged to v18 as this was proving a blocker.

Will try a minimal repro still though when I get some spare time 👍

**Comment by robahtou** [comment]
I'm a bit confused here (still new to resend). I made in issue here: https://github.com/resend/resend-node/issues/355
Not using tailwind. I simply installed the `resend` package with next canary & rc, made an action to call resend, `Resend error: TypeError: renderToStream is not a function` and the stacktrace shows `@react-email`. I'm confused about getting passed this error. What is the path foward? Fork this repo, make file changes, use the fork in our projects?

**Comment by gabrielmfern** [comment]
@robahtou Your issue is either #1486 or, if you are using React 19, it is that the Resend SDK hasn't yet updated the dependency on `@react-email/render`.

**Comment by LarsSalembier** [comment]
Is this going to get fixed or what is going on here? Seems trivial but I'm not that tech savvy and I can't be bothered to fork 5 different repos to fix this. As far as I can tell its just 2 lines of code that should be changed.

File: apps/web/package.json [hunk 5]
```diff
@@ -47,12 +47,12 @@
     "classnames": "2.5.1",
     "eslint": "^8",
     "eslint-config-custom": "workspace:*",
+    "postcss": "8.4.40",
+    "tailwindcss": "3.4.3",
```
Comment (vcapretz) [inline apps/web/package.json:51]: did this change position for some reason? can we leave it sorted? if we also remove the updated version from radix-ui/react-tooltip I believe this file would be unchanged and less things in the PR 😅

File: packages/button/src/__snapshots__/button.spec.tsx.snap [hunk 6]
```diff
@@ -2,6 +2,6 @@
...
 
-exports[`<Button> component > should allow users to overwrite style props 1`] = `"<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd"><!--$--><a style="line-height:150%;text-decoration:underline red;display:block;max-width:50%;mso-padding-alt:0px;padding:0px 0px 0px 0px" target="_blank"><span><!--[if mso]><i style="mso-font-width:NaN%;mso-text-raise:0" hidden></i><![endif]--></span><span style="max-width:100%;display:inline-block;line-height:120%;mso-padding-alt:0px;mso-text-raise:0"></span><span><!--[if mso]><i style="mso-font-width:NaN%" hidden>&#8203;</i><![endif]--></span></a><!--/$-->"`;
```
Comment (vcapretz) [inline packages/button/src/__snapshots__/button.spec.tsx.snap:7]: nice fix! no more `NaN` 😅

File: packages/button/src/button.tsx [hunk 7]
```diff
@@ -4,14 +4,30 @@ import { pxToPt } from "./utils/px-to-pt";
...
 export type ButtonProps = Readonly<React.ComponentPropsWithoutRef<"a">>;
 
+const maxFontWidth = 5;
+
+/**
+ * Computes a msoFontWidth \<= 5 and a count of space characters that,
+ * when applied, end up being as close to `expectedWidth` as possible.
+ */
 function computeFontWidthAndSpaceCount(expectedWidth: number) {
+  if (expectedWidth === 0) return [0, 0];
```
Comment (vcapretz) [inline packages/button/src/button.tsx:14]: can we use explicit blocks? 
```suggestion
  if (expectedWidth === 0) {
    return [0, 0];
  }
```

File: packages/create-email/CHANGELOG.md [hunk 8]
```diff
@@ -1,5 +1,15 @@
 # create-email
 
+## 0.0.35-canary.4
```
Comment (vcapretz) [inline packages/create-email/CHANGELOG.md:3]: should these be empty?

File: packages/link/src/link.tsx [hunk 9]
```diff
@@ -9,7 +9,7 @@ export const Link = React.forwardRef<HTMLAnchorElement, LinkProps>(
       ref={ref}
       style={{
         color: "#067df7",
-        textDecoration: "none",
+        textDecorationLine: "none",
```
Comment (vcapretz) [inline packages/link/src/link.tsx:12]: what's this change?

File: packages/react-email/package.json [hunk 10]
```diff
@@ -30,14 +30,14 @@
     "@babel/core": "7.24.5",
     "@babel/parser": "7.24.5",
     "chalk": "4.1.2",
-    "chokidar": "3.6.0",
+    "chokidar": "^4.0.1",
```
Comment (vcapretz) [inline packages/react-email/package.json:33]: are these updates required? if not, can we move them to a separate PR?

Comment (gabrielmfern) [reply packages/react-email/package.json]: This is a bit confusing, but because I merged instead of rebasing once, the files changed are a bit messy. Every time I do rebease it removes these extra changes, but for some reason they came back after? They should be gone now, just 77 instead of 150 files changed

File: apps/web/package.json (hunk 5 above)
Comment (gabrielmfern) [reply apps/web/package.json]: Done!

File: packages/tailwind/integrations/nextjs/src/app/page.tsx [hunk 11]
```diff
@@ -0,0 +1,7 @@
+import { renderAsync } from "@react-email/components";
+import { VercelInviteUserEmail } from "../../emails/vercel-invite-user";
+
+export default async function Home() {
+  const emailHTML = await renderAsync(<VercelInviteUserEmail />);
+  return <div dangerouslySetInnerHTML={{ __html: emailHTML }} />;
```
Comment (bukinoshita) [inline packages/tailwind/integrations/nextjs/src/app/page.tsx:6]:
```suggestion
  const emailHtml = await renderAsync(<VercelInviteUserEmail />);
  return <div dangerouslySetInnerHTML={{ __html: emailHtml }} />;
```

File: packages/tailwind/integrations/vite/package-lock.json
Comment (bukinoshita) [inline packages/tailwind/integrations/vite/package-lock.json:1]: Aren't we using `pnpm`?

File: packages/tailwind/src/utils/react/map-react-tree.ts [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+
+      if (
+        React.isValidElement<{ children?: React.ReactNode }>(processed) &&
+        (typeof processed.type === "function" ||
+          // @ts-expect-error - we know this is a component that may have a render function
+          processed.type.render)
+      ) {
+        const OriginalComponent =
+          typeof processed.type === "object"
+            ? // @ts-expect-error - we know this is a component with a render function
```
Comment (vcapretz) [inline packages/tailwind/src/utils/react/map-react-tree.ts:40]: what's the error you get if you don't include the expect error? it would be nice not to use it, perhaps some better assertion with type casting can already help

File: packages/tailwind/src/utils/react/is-component.ts [hunk 13]
```diff
@@ -0,0 +1,9 @@
+export const isComponent = (
+  element: React.ReactElement,
+): element is React.ReactElement<unknown, React.FC<unknown>> => {
+  return (
+    typeof element.type === "function" ||
+    // @ts-expect-error - we know this is a component that may have a render function
```
Comment (vcapretz) [inline packages/tailwind/src/utils/react/is-component.ts:6]: what's the error you get if you don't include the expect error? it would be nice not to use it, perhaps some better assertion with type casting can already help

File: packages/tailwind/src/utils/css/sanitize-declarations.ts [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+        const alpha = a === "1" || typeof a === "undefined" ? "" : `,${a}`;
+        return `rgb(${r},${g},${b}${alpha})`;
+      },
+    );
+
+    /*
+      We reset the regex's last index so that the state of the last time it ran doesn't intefere
+      with when it is used another time
+    */
+    regex.lastIndex = 0;
```
Comment (vcapretz) [inline packages/tailwind/src/utils/css/sanitize-declarations.ts:28]: can this be fixed by moving the regex declaration inside this function so we scope it to the function's lifetime?

File: packages/tailwind/src/utils/css/sanitize-declarations.ts [hunk 15]
```diff
@@ -0,0 +1,30 @@
+import type { AtRule, Root, Rule } from "postcss";
+
+const regex = /rgb\(\s*(\d+)\s*(\d+)\s*(\d+)(?:\s*\/\s*([\d%.]+))?\s*\)/g;
```
Comment (vcapretz) [inline packages/tailwind/src/utils/css/sanitize-declarations.ts:3]: can we find a better name that describes what regex is that? perhaps `declarationsRegex` or something

File: packages/tailwind/src/utils/css/resolve-all-css-variables.ts [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+    for (const [atRule, declarations] of declarationsForAtRules.entries()) {
+      const equivalentRule = createRule();
+      equivalentRule.selector = rule.selector;
+      equivalentRule.append(...declarations);
+
+      atRule.append(equivalentRule);
+    }
+  });
+
+  // Removes all variable definitions and then removes the rules that are empty
```
Comment (vcapretz) [inline packages/tailwind/src/utils/css/resolve-all-css-variables.ts:106]: I don't think this comment is needed

File: packages/tailwind/src/utils/css/resolve-all-css-variables.ts [hunk 17]
```diff
@@ -0,0 +1,118 @@
+/* eslint-disable @typescript-eslint/no-non-null-assertion */
```
Comment (vcapretz) [inline packages/tailwind/src/utils/css/resolve-all-css-variables.ts:1]: why are we disabling this?

File: packages/tailwind/integrations/nextjs/package-lock.json
Comment (vcapretz) [inline packages/tailwind/integrations/nextjs/package-lock.json:1]: since we use `pnpm` in the monorepo, I believe this should be a `pnpm-lock` file instead of the default `npm` one

File: packages/tailwind/src/utils/react/map-react-tree.ts (hunk 12 above)
Comment (gabrielmfern) [reply packages/tailwind/src/utils/react/map-react-tree.ts:40]: The error here is because the types for `type` aren't extensive enough to include the internal `render` from `forwardRef`. I remember trying to extend the type through separate type definitions, but things got too messy. We could do a type cast into something that is a bit more inclusive on this, for sure

File: packages/tailwind/src/hooks/use-tailwind.ts [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+import { resolveAllCSSVariables } from "../utils/css/resolve-all-css-variables";
+
+const tailwindAtRulesRoot = parse(
+  `
+  @tailwind base;
+  @tailwind components;
+`,
+).root();
+
+export function useTailwind(config: TailwindConfig) {
```
Comment (vcapretz) [inline packages/tailwind/src/hooks/use-tailwind.ts:21]: this seems to be more of a generic function rather than a React hook, if it's not a React hook can we find a different name so the `use..` doesn't create confusion?

File: packages/tailwind/src/hooks/use-tailwind.ts [hunk 19]
```diff
@@ -0,0 +1,47 @@
...
+  const tailwindContext = setupTailwindContext(config);
+
+  return {
+    // inspired by the tailwind LSP server
```
Comment (vcapretz) [inline packages/tailwind/src/hooks/use-tailwind.ts:25]:
```suggestion
```

File: packages/tailwind/src/utils/css/sanitize-declarations.ts (hunk 14 above)
Comment (gabrielmfern) [reply packages/tailwind/src/utils/css/sanitize-declarations.ts:28]: Definitely, but it might have a bit of a performance hit, since this function is called a bunch of times, might not be 100% worth the better readability

File: packages/tailwind/src/tailwind.tsx [hunk 20]
```diff
@@ -26,156 +27,93 @@ export interface TailwindProps {
...
-          : (element.type as React.FC);
-      const renderedComponent = component({
-        ...element.props,
-        ...propsToOverwrite,
-      });
-
-      if (React.isValidElement<EmailElementProps>(renderedComponent)) {
-        return processElement(renderedComponent);
+  if (hasNonInlineStylesToApply) {
+    let hasAppliedNonInlineStyles = false as boolean;
```
Comment (vcapretz) [inline packages/tailwind/src/tailwind.tsx:70]:
```suggestion
    let hasAppliedNonInlineStyles = false;
```
is the casting really needed?

File: packages/tailwind/src/utils/compatibility/convert-css-property-to-react-property.ts [hunk 21]
```diff
@@ -0,0 +1,21 @@
+import { fromDashCaseToCamelCase } from "../text/from-dash-case-to-camel-case";
+
+export const convertCssPropertyToReactProperty = (prop: string) => {
+  let modifiedProp = prop;
+
+  modifiedProp = modifiedProp.toLowerCase();
```
Comment (vcapretz) [inline packages/tailwind/src/utils/compatibility/convert-css-property-to-react-property.ts:6]:
```suggestion
  const modifiedProp = prop.toLowerCase();
```

File: packages/tailwind/src/utils/compatibility/convert-css-property-to-react-property.ts [hunk 22]
```diff
@@ -0,0 +1,21 @@
...
+  if (modifiedProp === "float") {
+    return "cssFloat";
+  }
+
+  if (modifiedProp.startsWith("--")) {
+    return modifiedProp;
+  }
+
+  if (modifiedProp.startsWith("-ms-")) {
+    modifiedProp = modifiedProp.slice(1);
```
Comment (vcapretz) [inline packages/tailwind/src/utils/compatibility/convert-css-property-to-react-property.ts:17]:
```suggestion
    return fromDashCaseToCamelCase(modifiedProp.slice(1));
```

File: packages/tailwind/src/utils/css/resolve-all-css-variables.ts (hunk 17 above)
Comment (gabrielmfern) [reply packages/tailwind/src/utils/css/resolve-all-css-variables.ts:1]: Because of this portion here,

https://github.com/resend/react-email/blob/3ecc61b0c2cc443ebe42789a9003972e3aa51739/packages/tailwind/src/utils/css/resolve-all-css-variables.ts#L70-L73

we could also do a single get, and then check if the resulting value is not undefined

File: packages/tailwind/src/utils/css/make-inline-styles-for.ts [hunk 23]
```diff
@@ -0,0 +1,61 @@
...
+import { convertCssPropertyToReactProperty } from "../compatibility/convert-css-property-to-react-property";
+
+const walkInlinableRules = (root: Root, callback: (rule: Rule) => void) => {
+  root.walkRules((rule) => {
+    if (rule.parent?.type === "atrule") {
+      return;
+    }
+
+    selectorParser((selector) => {
+      let hasPseudoSelectors = false as boolean;
```
Comment (vcapretz) [inline packages/tailwind/src/utils/css/make-inline-styles-for.ts:13]:
```suggestion
      let hasPseudoSelectors = false;
```

File: packages/tailwind/integrations/vite/package-lock.json
Comment (gabrielmfern) [reply packages/tailwind/integrations/vite/package-lock.json:1]: Yup, but the idea with these is to have something as similar to a user environment as possible, so I ended up using npm directly instead of pnpm - because it would force us into the workspace, and make it less accurate. You guys have any better ideas?

File: packages/tailwind/integrations/vite/package.json [hunk 24]
```diff
@@ -0,0 +1,33 @@
...
+    "build": "tsc && vite build",
+    "lint": "eslint . --ext ts,tsx --report-unused-disable-directives --max-warnings 0",
+    "preinstall:tailwind": "cd ../../ && turbo build && yalc installations clean @react-email/tailwind && yalc publish",
+    "preinstall": "npm run preinstall:tailwind && yalc add @react-email/tailwind",
+    "preview": "vite preview"
+  },
+  "dependencies": {
+    "react": "^18.2.0",
+    "@react-email/components": "^0.0.15",
+    "@react-email/tailwind": "file:.yalc/@react-email/tailwind",
```
Comment (bukinoshita) [inline packages/tailwind/integrations/vite/package.json:17]: Is this correct?

File: packages/tailwind/src/hooks/use-clone-element-with-inlined-styles.ts [hunk 25]
```diff
@@ -0,0 +1,78 @@
...
+import { makeInlineStylesFor } from "../utils/css/make-inline-styles-for";
+import { sanitizeClassName } from "../utils/compatibility/sanitize-class-name";
+import { isComponent } from "../utils/react/is-component";
+import type { useTailwind } from "./use-tailwind";
+
+export const useCloneElementWithInlinedStyles = (
+  tailwind: ReturnType<typeof useTailwind>,
+) => {
+  return (element: React.ReactElement<EmailElementProps>) => {
+    const propsToOverwrite = {} as Partial<EmailElementProps>;
```
Comment (bukinoshita) [inline packages/tailwind/src/hooks/use-clone-element-with-inlined-styles.ts:15]: Do we need to cast?

File: packages/tailwind/src/tailwind.tsx [hunk 26]
```diff
@@ -26,156 +27,93 @@ export interface TailwindProps {
...
-          styleElement,
-        );
+  const tailwind = useTailwind(config ?? {});
+
+  const cloneElementWithInlinedStyles =
+    useCloneElementWithInlinedStyles(tailwind);
+
+  const nonInlineStylesRootToApply = new Root();
+  let mediaQueryClassesForAllElement: string[] = [];
+  let hasNonInlineStylesToApply = false as boolean;
```
Comment (bukinoshita) [inline packages/tailwind/src/tailwind.tsx:44]: Do we need to cast?

File: packages/tailwind/src/tailwind.tsx [hunk 27]
```diff
@@ -26,156 +27,93 @@ export interface TailwindProps {
...
+
+  let mappedChildren = mapReactTree(children, (node) => {
+    if (React.isValidElement<EmailElementProps>(node)) {
+      const element = node;
```
Comment (bukinoshita) [inline packages/tailwind/src/tailwind.tsx:48]: Shouldn't we just use `node` directly instead of assigning to a variable?

File: packages/tailwind/src/tailwind.tsx (hunk 26 above)
Comment (gabrielmfern) [reply packages/tailwind/src/tailwind.tsx:44]: The problem here is linting. If we use a proper type for the variable eslint complains about types that can be inferred, and if we don't cast, eslint complains that the condition down below is unnecessary, when it actually is necessary.

https://github.com/resend/react-email/blob/560c789acca66d1a7bc68529443efdc69a87cd98/packages/tailwind/src/tailwind.tsx#L66-L68

The best option I found to not disable linting here was to cast the type.

File: packages/tailwind/integrations/vite/package.json (hunk 24 above)
Comment (gabrielmfern) [reply packages/tailwind/integrations/vite/package.json:17]: Yup, if we test importing from the workspace directly the testing won't be as conclusive as this one, a lot of issues would get by without noticing from my experience.

File: packages/tailwind/src/utils/css/make-inline-styles-for.ts (hunk 23 above)
Comment (gabrielmfern) [reply packages/tailwind/src/utils/css/make-inline-styles-for.ts:13]: The problem here is linting. If we use a proper type for the variable eslint complains about types that can be inferred, and if we don't cast, eslint complains that the condition down below is unnecessary, when it actually is necessary.

https://github.com/resend/react-email/blob/560c789acca66d1a7bc68529443efdc69a87cd98/packages/tailwind/src/utils/css/make-inline-styles-for.ts#L18-L20

The best option I found to not disable linting here was to cast the type.

File: packages/tailwind/integrations/nextjs/package-lock.json
Comment (gabrielmfern) [reply packages/tailwind/integrations/nextjs/package-lock.json:1]: The reason for these is to simulate a user environment as closely possible, so I ended up using npm directly instead of pnpm — because it would force us into the workspace, and make it less accurate. You guys have any better ideas?

**Review by vcapretz — APPROVED** [review]

File: packages/tailwind/src/utils/react/is-component.ts (hunk 13 above)
Comment (gabrielmfern) [reply packages/tailwind/src/utils/react/is-component.ts:6]: Initially I tried defining a few lines of code to get around this, but the amount of code that's going to need changing is a lot compared to get around the expect error. 😞

The issue is that React doesn't include the `render` in their types as it isn't something for user land, mostly.
